src/c-writer.cc

```diff
@@ -2712,6 +2940,101 @@ void CWriter::Write(const Func& func) {
+  WriteTailCallAsserts(func.decl.sig);
+  Write(ModuleInstanceTypeName(), "* instance = *instance_ptr;", Newline());
+
+  std::vector<std::string> index_to_name;
+  MakeTypeBindingReverseMapping(func_->GetNumParamsAndLocals(), func_->bindings,
+                                &index_to_name);
+  if (func_->GetNumParams() == 1) {
+    Write(func_->GetParamType(0), " ", DefineParamName(index_to_name[0]),
+          " = *(", func_->GetParamType(0), "*)tail_call_stack;", Newline());
+  } else if (func_->GetNumParams() > 1) {
+    for (Type type : {Type::I32, Type::I64, Type::F32, Type::F64, Type::V128,
+                      Type::FuncRef, Type::ExternRef}) {
+      Index param_index = 0;
+      size_t count = 0;
+      for (Type param_type : func_->decl.sig.param_types) {
+        if (param_type == type) {
+          if (count == 0) {
+            Write(type, " ");
+            Indent(4);
+          } else {
+            Write(", ");
+            if ((count % 8) == 0)
+              Write(Newline());
+          }
+
+          Write(DefineParamName(index_to_name[param_index]));
+          ++count;
+        }
+        ++param_index;
+      }
+      if (count != 0) {
+        Dedent(4);
+        Write(";", Newline());
+      }
+    }
+    Write(OpenBrace(), func_->decl.sig.param_types, " *tmp = tail_call_stack;",
+          Newline());
+    Unspill(func_->decl.sig.param_types, true,
+            [&](auto i) { return ParamName(index_to_name[i]); });
+    Write(CloseBrace(), Newline());
+  }
+
+  WriteLocals(index_to_name);
+
+  PushFuncSection();
+
+  std::string label = DefineLabelName(kImplicitFuncLabel);
+  ResetTypeStack(0);
+  std::string empty;  // Must not be temporary, since address is taken by Label.
+  PushLabel(LabelType::Func, empty, func.decl.sig);
+  Write(func.exprs, LabelDecl(label));
+  PopLabel();
+  ResetTypeStack(0);
+  PushTypes(func.decl.sig.result_types);
+
+  // Return the top of the stack implicitly.
+  Index num_results = func.GetNumResults();
+  if (num_results == 1) {
+    Write("wasm_rt_memcpy(tail_call_stack, &", StackVar(0), ", sizeof(",
+          StackVar(0), "));", Newline());
+  } else if (num_results > 1) {
+    Write(OpenBrace(), func.decl.sig.result_types, " *tmp = tail_call_stack;",
+          Newline());
+    Spill(func.decl.sig.result_types, true);
+    Write(CloseBrace(), Newline());
+  }
+  Write("next->fn = NULL;", Newline());
+
+  stream_ = prev_stream;
+
+  for (size_t i = 0; i < func_sections_.size(); ++i) {
+    auto& [condition, stream] = func_sections_.at(i);
+    std::unique_ptr<OutputBuffer> buf = stream.ReleaseOutputBuffer();
+    if (condition.empty() || func_includes_.count(condition)) {
+      stream_->WriteData(buf->data.data(), buf->data.size());
+    }
+
+    if (i == 0) {
+      WriteStackVarDeclarations();  // these come immediately after section #0
+                                    // (return type/name/params/locals)
+    }
+  }
+
+  Write(CloseBrace(), Newline());
```

Comment:
It would be nice to dedup this code from `CWriter::Write(const Func& func)` if they are the same

Reply:
Done in 934b6695a314d54514f62a3f8989ffafeb3e3874